{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get the sequences from a gtf annotation and the corresponding fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from mrna_dataloader_general import MRNA_halflife_dl\n",
    "\n",
    "#This files can be downloaded from the gencode website (https://www.gencodegenes.org/human/release_25.html)\n",
    "gtf_path =  '../../data/gencode.v25.annotation.gtf.gz'\n",
    "fasta_path = '../../data/hg38.fa'\n",
    "\n",
    "mrna_hl_dl = MRNA_halflife_dl(gtf_file_path=gtf_path, fasta_file_path=fasta_path,\n",
    "                              mrna_region='cds', get_codon_number=True, u_aug_search=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2579812</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15887</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>MT-TT-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2579813</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15887</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>MT-TT-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001544475.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2579814</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15955</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2579815</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15955</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>MT-TP-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2579816</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15955</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>MT-TP-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001544473.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2579817 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome   Source     Feature  Start    End Score Strand Frame  \\\n",
       "0             chr1   HAVANA        gene  11868  14409     .      +     .   \n",
       "1             chr1   HAVANA  transcript  11868  14409     .      +     .   \n",
       "2             chr1   HAVANA        exon  11868  12227     .      +     .   \n",
       "3             chr1   HAVANA        exon  12612  12721     .      +     .   \n",
       "4             chr1   HAVANA        exon  13220  14409     .      +     .   \n",
       "...            ...      ...         ...    ...    ...   ...    ...   ...   \n",
       "2579812       chrM  ENSEMBL  transcript  15887  15953     .      +     .   \n",
       "2579813       chrM  ENSEMBL        exon  15887  15953     .      +     .   \n",
       "2579814       chrM  ENSEMBL        gene  15955  16023     .      -     .   \n",
       "2579815       chrM  ENSEMBL  transcript  15955  16023     .      -     .   \n",
       "2579816       chrM  ENSEMBL        exon  15955  16023     .      -     .   \n",
       "\n",
       "                   gene_id                           gene_type  ...  \\\n",
       "0        ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "1        ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "2        ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "3        ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "4        ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "...                    ...                                 ...  ...   \n",
       "2579812  ENSG00000210195.2                             Mt_tRNA  ...   \n",
       "2579813  ENSG00000210195.2                             Mt_tRNA  ...   \n",
       "2579814  ENSG00000210196.2                             Mt_tRNA  ...   \n",
       "2579815  ENSG00000210196.2                             Mt_tRNA  ...   \n",
       "2579816  ENSG00000210196.2                             Mt_tRNA  ...   \n",
       "\n",
       "        transcript_status transcript_name transcript_support_level    tag  \\\n",
       "0                     NaN             NaN                      NaN    NaN   \n",
       "1                   KNOWN     DDX11L1-002                        1  basic   \n",
       "2                   KNOWN     DDX11L1-002                        1  basic   \n",
       "3                   KNOWN     DDX11L1-002                        1  basic   \n",
       "4                   KNOWN     DDX11L1-002                        1  basic   \n",
       "...                   ...             ...                      ...    ...   \n",
       "2579812             KNOWN       MT-TT-201                       NA  basic   \n",
       "2579813             KNOWN       MT-TT-201                       NA  basic   \n",
       "2579814               NaN             NaN                      NaN    NaN   \n",
       "2579815             KNOWN       MT-TP-201                       NA  basic   \n",
       "2579816             KNOWN       MT-TP-201                       NA  basic   \n",
       "\n",
       "            havana_transcript exon_number            exon_id  ont protein_id  \\\n",
       "0                         NaN         NaN                NaN  NaN        NaN   \n",
       "1        OTTHUMT00000362751.1         NaN                NaN  NaN        NaN   \n",
       "2        OTTHUMT00000362751.1           1  ENSE00002234944.1  NaN        NaN   \n",
       "3        OTTHUMT00000362751.1           2  ENSE00003582793.1  NaN        NaN   \n",
       "4        OTTHUMT00000362751.1           3  ENSE00002312635.1  NaN        NaN   \n",
       "...                       ...         ...                ...  ...        ...   \n",
       "2579812                   NaN         NaN                NaN  NaN        NaN   \n",
       "2579813                   NaN           1  ENSE00001544475.2  NaN        NaN   \n",
       "2579814                   NaN         NaN                NaN  NaN        NaN   \n",
       "2579815                   NaN         NaN                NaN  NaN        NaN   \n",
       "2579816                   NaN           1  ENSE00001544473.2  NaN        NaN   \n",
       "\n",
       "        ccdsid  \n",
       "0          NaN  \n",
       "1          NaN  \n",
       "2          NaN  \n",
       "3          NaN  \n",
       "4          NaN  \n",
       "...        ...  \n",
       "2579812    NaN  \n",
       "2579813    NaN  \n",
       "2579814    NaN  \n",
       "2579815    NaN  \n",
       "2579816    NaN  \n",
       "\n",
       "[2579817 rows x 26 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mrna_hl_dl.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chromosome': 'chr1',\n",
       " 'gene_id': 'ENSG00000279457.3',\n",
       " 'transcript_id': 'ENST00000623083.3',\n",
       " 'cds': 'ATGACTCCTGTGAGGATGCAGCACTCCCTGGCAGGTCAGACCTATGCCGTGCCCCTCATCCAGCCAGACCTGCGGCGAGAGGAGGCCGTCCAGCAGATGGCGGATGCCCTGCAGTACCTGCAGAAGGTCTCTGGAGACATCTTCAGCAGGATCTCCCAGCAGGTAGAGCAGAGCCGGAGCCAGGTGCAGGCCATTGGAGAGAAGGTCTCCTTGGCCCAGGCCAAGATTGAGAAGATCAAGGGCAGCAAGAAGGCCATCAAGGTGTTCTCCAGTGCCAAGTACCCTGCTCCAGGGCGCCTGCAGGAATATGGCTCCATCTTCACGGGCGCCCAGGACCCTGGCCTGCAGAGACGCCCCCACAGGATCCAGAGCAAGCACCGCCCCCTGGACGAGCGGGCCCTGCAGGAGAAGCTGAAGGACTTTCCTGTGTGCGTGAGCACCAAGCCGGAGCCCGAGGACGATGCAGAAGAGGGACTTGGGGGTCTTCCCAGCAACATCAGCTCTGTCAGCTCCTTGCTGCTCTTCAACACCACCGAGAACCTGTATAAGAAGTATGTCTTCCTGGACCCCCTGGCTGGTGCTGTAACAAAGACCCATGTGATGCTGGGGGCAGAGACAGAGGAGAAGCTGTTTGATGCCCCCTTGTCCATCAGCAAGAGAGAGCAGCTGGAACAGCAGGTCCCAGAGAACTACTTCTATGTGCCAGACCTGGGCCAGGTGCCTGAGATTGATGTTCCATCCTACCTGCCTGACCTGTCCGGCATTGCCAACGACCTCATGTACATTGCCGACCTGGGCCCCGGCATTGCCCCCTCTGCCCCTGGCACCATTCCAGAACTGCCCACCTTCCACACTGAGGTAGCCGAGCCTCTCAAGGTAGACCTACAAGATGGGGTACTAACACCACCCCCACCGCCCCCACCACCACCCCCAGCTCCTGAGGTGCTGGCCAGTGCACCCCCACTCCCACCCTCAACCGCGGCCCCTGTAGGCCAAGGCGCCAGGCAGGACGACAGCAGCAGCAGCGCGTCTCCTTCAGTCCAGGGAGCTCCCAGGGAAGTGGTCGACCCCTCCGGTGGCCGGGCCACTCTGCTAGAGTCCATCCGCCAAGCTGGGGGCATCGGCAAGGCCAAGCTGCGCAGCATGAAGGAGCGAAAGCTGGAGAAGAAGAAGCAGAAGGAGCAGGAGCAAGTGAGAGCCACGAGCCAAGGTGGGCACTTGATGTCGGATCTCTTCAACAAGCTGGTCATGAGGCGCAAGGGCATCTCTGGGAAAGGACCTGGGGCTGGTGAGGGGCCCGGAGGAGCCTTTGCCCGCGTGTCAGACTCCATCCCTCCTCTGCCGCCACCGCAGCAGCCACAGGCAGAGGAGGACGAGGACGACTGGGAATCC',\n",
       " 'non_AUG': False,\n",
       " 'codon_numbers':      codon number\n",
       " AAA           1.0\n",
       " AAC           6.0\n",
       " AAG          29.0\n",
       " AAT           0.0\n",
       " ACA           3.0\n",
       " ..            ...\n",
       " TGT           0.0\n",
       " TTA           0.0\n",
       " TTC           8.0\n",
       " TTG           4.0\n",
       " TTT           3.0\n",
       " \n",
       " [64 rows x 1 columns]}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mrna_hl_dl[3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The sequences are stored into a pickle file, in this case \"loaded_seqs_gencode25\". The other stored sequences were saved under the name \"loaded_data_gencode19\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading_data...\n",
      "0.000000 per cent completed\n",
      "1.065008 per cent completed\n",
      "2.130016 per cent completed\n",
      "3.195024 per cent completed\n",
      "4.260032 per cent completed\n",
      "5.325040 per cent completed\n",
      "6.390049 per cent completed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3331, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-3-0801ceed91e6>\", line 1, in <module>\n",
      "    data = mrna_hl_dl.load_data(save_path=\"loaded_seqs_gencode25\")\n",
      "  File \"/data/nasif12/home_if12/tomazdas/Projects/mrna-halflife/data_utils/mrna_dataloader_general.py\", line 185, in load_data\n",
      "    for transcript_i, mrna in enumerate(self):\n",
      "  File \"/data/nasif12/home_if12/tomazdas/Projects/mrna-halflife/data_utils/mrna_dataloader_general.py\", line 155, in __getitem__\n",
      "    data['cds'], data['non_AUG'] = self.cds_fetcher.get_seq(transcript)\n",
      "  File \"/data/nasif12/home_if12/tomazdas/Projects/mrna-halflife/data_utils/mrna_dataloader_general.py\", line 54, in get_seq\n",
      "    exons, strand = self.get_cds_exons(transcript_id)\n",
      "  File \"/data/nasif12/home_if12/tomazdas/Projects/mrna-halflife/data_utils/mrna_dataloader_general.py\", line 34, in get_cds_exons\n",
      "    cds_exons = self.cds.loc[transcript_id]\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/site-packages/pandas/core/indexing.py\", line 1768, in __getitem__\n",
      "    return self._getitem_axis(maybe_callable, axis=axis)\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/site-packages/pandas/core/indexing.py\", line 1965, in _getitem_axis\n",
      "    return self._get_label(key, axis=axis)\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/site-packages/pandas/core/indexing.py\", line 625, in _get_label\n",
      "    return self.obj._xs(label, axis=axis)\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/site-packages/pandas/core/generic.py\", line 3537, in xs\n",
      "    loc = self.index.get_loc(key)\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/site-packages/pandas/core/indexes/base.py\", line 2646, in get_loc\n",
      "    return self._engine.get_loc(key)\n",
      "  File \"pandas/_libs/index.pyx\", line 111, in pandas._libs.index.IndexEngine.get_loc\n",
      "  File \"pandas/_libs/index.pyx\", line 133, in pandas._libs.index.IndexEngine.get_loc\n",
      "  File \"pandas/_libs/index.pyx\", line 163, in pandas._libs.index.IndexEngine._get_loc_duplicates\n",
      "  File \"pandas/_libs/index.pyx\", line 172, in pandas._libs.index.IndexEngine._maybe_get_bool_indexer\n",
      "  File \"<__array_function__ internals>\", line 2, in where\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2044, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'KeyboardInterrupt' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1148, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 316, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 350, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/inspect.py\", line 739, in getmodule\n",
      "    f = getabsfile(module)\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/inspect.py\", line 708, in getabsfile\n",
      "    _filename = getsourcefile(object) or getfile(object)\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/inspect.py\", line 693, in getsourcefile\n",
      "    if os.path.exists(filename):\n",
      "  File \"/opt/modules/i12g/anaconda/envs/mrna_halflife2/lib/python3.7/genericpath.py\", line 19, in exists\n",
      "    os.stat(path)\n",
      "KeyboardInterrupt\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m"
     ]
    }
   ],
   "source": [
    "data = mrna_hl_dl.load_data(save_path=\"loaded_seqs_gencode25\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:anaconda-mrna_halflife2]",
   "language": "python",
   "name": "conda-env-anaconda-mrna_halflife2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
